contributingGuides/STYLE.md

```diff
@@ -10,13 +68,497 @@ We use Prettier to automatically style our code.
+
+```ts
+// Array<T>
+const a: Array<string | number> = ["a", "b"];
+const b: Array<{ prop: string }> = [{ prop: "a" }];
+const c: Array<() => void> = [() => {}];
+
+// T[]
+const d: MyType[] = ["a", "b"];
+const e: string[] = ["a", "b"];
+const f: readonly string[] = ["a", "b"];
+```
+
+### `@ts-ignore`
+
+Do not use `@ts-ignore` or its variant `@ts-nocheck` to suppress warnings and errors.
+
+### Type Inference
+
+When possible, allow the compiler to infer type of variables.
+
+```ts
+// BAD
+const foo: string = "foo";
+const [counter, setCounter] = useState<number>(0);
+
+// GOOD
+const foo = "foo";
+const [counter, setCounter] = useState(0);
+const [username, setUsername] = useState<string | undefined>(undefined); // Username is a union type of string and undefined, and its type cannot be inferred from the default value of undefined
+```
+
+For function return types, default to always typing them unless a function is simple enough to reason about its return type.
+
+> Why? Explicit return type helps catch errors when implementation of the function changes. It also makes it easy to read code even when TypeScript intellisense is not provided.
+
+```ts
+function simpleFunction(name: string) {
+    return `hello, ${name}`;
+}
+
+function complicatedFunction(name: string): boolean {
+// ... some complex logic here ...
+    return foo;
+}
+```
+
+### Utility Types
+
+Use types from [TypeScript utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) and [`type-fest`](https://github.com/sindresorhus/type-fest) when possible.
+
+```ts
+type Foo = {
+    bar: string;
+};
+
+// BAD
+type ReadOnlyFoo = {
+    readonly [Property in keyof Foo]: Foo[Property];
+};
+
+// GOOD
+type ReadOnlyFoo = Readonly<Foo>;
+```
+
+### `object` type
+
+Don't use `object` type. eslint: [`@typescript-eslint/ban-types`](https://typescript-eslint.io/rules/ban-types/)
+
+> Why? `object` refers to "any non-primitive type," not "any object". Typing "any non-primitive value" is not commonly needed.
+
+```ts
+// BAD
+const foo: object = [1, 2, 3]; // TypeScript does not error
+```
+
+If you know that the type of data is an object but don't know what properties or values it has beforehand, use `Record<string, unknown>`.
+
+> Even though `string` is specified as a key, `Record<string, unknown>` type can still accepts objects whose keys are numbers. This is because numbers are converted to strings when used as an object index. Note that you cannot use [symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) for `Record<string, unknown>`.
+
+```ts
+function logObject(object: Record<string, unknown>) {
+    for (const [key, value] of Object.entries(object)) {
+        console.log(`${key}: ${value}`);
+    }
+}
+```
+
+### Prop Types
+
+Don't use `ComponentProps` to grab a component's prop types. Go to the source file for the component and export prop types from there. Import and use the exported prop types.
```

Comment:
I'm actually not sure why we decided this, so maybe we should add a `why` clause here?

Reply:
I'm not 100% sure either, as it was copied from `TS_STYLE.md`. I think it might be mostly for consistency, I have also seen some threads with people complaining about `ComponentProps` not always working, e.g. for class components with default props. Any ideas what we should put in the **why** section? 

CC: @fabioh8010 @blazejkustra

Reply:
Yes, since we have our own props defined in the codebase is always easy to just export and use it when required. `ComponentProps` (and its friends `ComponentPropsWithoutRef`, ComponentPropsWithRef`) should be used only for specific situations or generic components.

Reply:
Thanks guys, changes applied!